### PR TITLE
feature: ssl: add FFI functions to parse cert/priv key to cdata instead of DER

### PR DESCRIPTION
**Pull request by ghedo**
Adds Lua functions for https://github.com/openresty/lua-nginx-module/pull/764

### CODE REVIEW

**Comment by ghedo** [comment]
It seems that `ssl.t` plans too many tests:
https://github.com/openresty/lua-resty-core/blob/master/t/ssl.t#L12
The `+ 1` doesn't seem to be needed, and it causes the test suite to fail. Maybe I'm doing something wrong...

**Comment by ghedo** [comment]
Also, have you ever thought about running the OpenResty repos (not just lua-resty-core) builds and tests on Travis CI? It would make it easier to see which PRs pass the test suite. Though it will mean building nginx+lua-nginx-module every time.

**Comment by agentzh** [comment]
@ghedo The tests on the current git master are passing completely for me:

```
t/balancer.t ........ 49/154 nginx: [warn] load balancing method redefined in /home/agentzh/git/lua-resty-core/t/servroot/conf/nginx.conf:34
t/balancer.t ........ ok
t/count.t ........... ok
t/ctx.t ............. ok
t/decode-base64.t ... ok
t/encode-base64.t ... ok
t/exit.t ............ ok
t/md5.t ............. ok
t/md5_bin.t ......... ok
t/misc.t ............ ok
t/ocsp.t ............ 1/230 WARNING: TEST 8: create OCSP request (empty string cert chain) - 2016/05/13 14:26:01 [crit] 1712#0: *3 SSL_do_handshake() failed (SSL: error:0D07207B:asn1 encoding routines:ASN1_get_object:header too long error:1408A179:SSL routines:ssl3_get_client_hello:cert cb error) while SSL handshaking, client: unix:, server: unix:/home/agentzh/git/lua-resty-core/t/servroot/html/nginx.sock at /home/agentzh/git/lua-resty-core/../test-nginx/lib/Test/Nginx/Socket.pm line 1192.
WARNING: TEST 8: create OCSP request (empty string cert chain) - 2016/05/13 14:26:01 [crit] 1712#0: *6 SSL_do_handshake() failed (SSL: error:0D07207B:asn1 encoding routines:ASN1_get_object:header too long error:1408A179:SSL routines:ssl3_get_client_hello:cert cb error) while SSL handshaking, client: unix:, server: unix:/home/agentzh/git/lua-resty-core/t/servroot/html/nginx.sock at /home/agentzh/git/lua-resty-core/../test-nginx/lib/Test/Nginx/Socket.pm line 1192.
t/ocsp.t ............ 107/230 WARNING: TEST 15: fail to validate OCSP response - OCSP response returns revoked status - 2016/05/13 14:26:02 [crit] 1842#0: *3 SSL_do_handshake() failed (SSL: error:1408A179:SSL routines:ssl3_get_client_hello:cert cb error) while SSL handshaking, client: unix:, server: unix:/home/agentzh/git/lua-resty-core/t/servroot/html/nginx.sock at /home/agentzh/git/lua-resty-core/../test-nginx/lib/Test/Nginx/Socket.pm line 1192.
WARNING: TEST 15: fail to validate OCSP response - OCSP response returns revoked status - 2016/05/13 14:26:02 [crit] 1842#0: *6 SSL_do_handshake() failed (SSL: error:1408A179:SSL routines:ssl3_get_client_hello:cert cb error) while SSL handshaking, client: unix:, server: unix:/home/agentzh/git/lua-resty-core/t/servroot/html/nginx.sock at /home/agentzh/git/lua-resty-core/../test-nginx/lib/Test/Nginx/Socket.pm line 1192.
t/ocsp.t ............ ok
t/re-find.t ......... ok
t/re-match.t ........ ok
t/re-sub.t .......... ok
t/request.t ......... ok
t/response.t ........ ok
t/semaphore.t ....... 10/246 nginx: [warn] [lua] init_by_lua:7: sema created: table: 0x41873ac0
t/semaphore.t ....... 185/246 nginx: [warn] [lua] init_by_lua:7: sema created: table: 0x41084080
t/semaphore.t ....... ok
t/sha1_bin.t ........ ok
t/shdict.t .......... ok
t/ssl.t ............. ok
t/status.t .......... ok
t/time.t ............ ok
t/uri.t ............. ok
t/var.t ............. ok
t/worker.t .......... ok
t/worker_count_5.t .. ok
All tests successful.
Files=25, Tests=3436, 36 wallclock secs ( 0.57 usr  0.05 sys +  8.71 cusr  4.99 csys = 14.32 CPU)
Result: PASS
```

Maybe you've got something wrong. Please ensure that you are using the latest git master version of Test::Nginx.

Regarding Travis CI integration, yes, we can do that. Patches welcome. Now I'm manually ensuring that the master always passes all the tests. More thorough testing is conducted on my EC2 test cluster:

http://qa.openresty.org

**Comment by ghedo** [comment]
Retried on Linux with Test::Nginx from master and the plan-related error is gone. I do get a bunch of other failures though: https://gist.github.com/ghedo/e11bb15b99b90553021094613851d7ac

This is with lua-nginx-module and nginx master (not the openresty bundle), so that may be related.

**Comment by ghedo** [comment]
In any case it seems unrelated to my changes :)

**Comment by ghedo** [comment]
Actually, the command above still used the cpanm-installed version of Test::Nginx, I retried with the proper command after removing Test::Nginx from the load path and I still got the same errors.

**Comment by agentzh** [comment]
@ghedo Well, just use `prove -I /path/to/test-nginx/lib -r t` to run the tests to ensure you are using the latest git master of test-nginx.

**Comment by agentzh** [comment]
@ghedo I'll look into this PR after OpenResty 1.9.15.1 is out. Thanks for your hard work on this!

File: lib/ngx/ssl.md [hunk 1]
```diff
@@ -308,6 +312,68 @@ This function can be called in whatever contexts where downstream https is used.
...
+
+Converts the PEM-formated SSL certificate chain data into an opaque cdata pointer (for later uses
+in the [set_cert](#set_cert)
+function, for example).
+
+In case of failures, returns `nil` and a string describing the error.
+
+You can always use libraries like [lua-resty-lrucache](https://github.com/openresty/lua-resty-lrucache#readme)
+and/or ngx_lua APIs like [lua_shared_dict](https://github.com/openresty/lua-nginx-module#lua_shared_dict)
+to do the caching of the cdata results, for example.
```
Comment (agentzh) [inline lib/ngx/ssl.md]: I don't think lua_shared_dict can cache the cdata returned directly. We need some kind of serialization here. Thoughts?

**Comment by agentzh** [comment]
@ghedo Merged into master with some minor changes. Many thanks!

**Pull request closed without merging**